integration_test.go

```diff
@@ -0,0 +1,279 @@
+
+	qBody, err := ioutil.ReadAll(qResp.Body)
+	qResp.Body.Close()
+	require.NoError(t, err, "read from quick request body")
+	require.NotContains(t, string(qBody), fmt.Sprintf("%d", ppid), "quick response contained wrong PID")
+	require.Contains(t, string(qBody), fmt.Sprintf("%d", cpid), "quick response did not contain correct PID")
+
+	// Wait for long request to finish, which should be responded to from
+	// the parent integration test server
+	lReqWait.Wait()
+	require.NoError(t, lReqErr, "sending long request to parent server")
+	require.Equal(t, http.StatusOK, lResp.StatusCode, "http response code from long running request")
+	lBody, err := ioutil.ReadAll(lResp.Body)
+	require.NoError(t, err, "read from long request body")
+	require.Contains(t, string(lBody), fmt.Sprintf("%d", ppid), "long response did not contain correct PID")
+	time.Sleep(5 * time.Second)
+	require.Error(t, intServer.Process.Signal(syscall.Signal(0)), "Expected error when signalling stopped parent process, process has not exited")
+
+}
+
+func sendIntegrationTestRequest(sleepDuration time.Duration) (*http.Response, error) {
+	reqURL := fmt.Sprintf("http://localhost:%s%s?duration=%s", intSrvPort, intSrvPath, sleepDuration)
+	timeout := int(sleepDuration.Seconds() + 2) // add two seconds to the duration of the call to get the http timeout
+
+	client := http.Client{
+		Timeout: time.Duration(timeout) * time.Second,
+	}
+	return client.Get(reqURL)
+}
+
+func startIntegrationServer(intServer *exec.Cmd, messages chan string) error {
+	outPipe, err := intServer.StdoutPipe()
+	if err != nil {
+		close(messages)
+		return err // note process not started yet so we shouldn't have to kill intserver
+	}
+	if err := intServer.Start(); err != nil {
+		close(messages)
+		return err
+	}
+	// Forward relevant messages from stdout back to the test
+	go watchStdout(outPipe, messages)
+	return nil
+}
+
+func teardownServer(pid int) {
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		fmt.Printf("Could not find process with PID %d to clean up\n", pid)
+	}
+	if err := proc.Kill(); err != nil {
+		if !strings.Contains(err.Error(), "process already finished") {
+			fmt.Printf(
+				"Could not kill process with PID %d with err %v. Manual clean up may be required\n",
+				pid, err)
+		}
+	}
+}
+
+func waitForMessage(msgs chan string, timeout time.Duration) (string, error) {
+	to := time.NewTimer(timeout)
+	select {
+	case <-to.C:
+		return "", fmt.Errorf("timeout exceeded")
+	case msg, open := <-msgs:
+		if !open {
+			return "", fmt.Errorf("unexpected close of messages channel")
+		}
+		return msg, nil
+	}
+}
+
+func waitForServer(uri string) error {
+	reqURL, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return err
+	}
+	client := http.DefaultClient
+	ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
+	defer cancel()
+	for {
+		cctx, ccancel := context.WithTimeout(ctx, 10*time.Millisecond)
+		defer ccancel()
+		req := &http.Request{
```

Comment:
loop invarient

Reply:
I wasn't sure whether I could re-use the request, or just the client, will move the request outside the loop.